atomic_reactor/plugins/exit_koji_import.py

```diff
@@ -123,21 +127,107 @@ def __init__(self, tasker, workflow, kojihub=None, url=None,
+        output = Output(file=open(saved_image), metadata=metadata)
+
+        return metadata, output
+
+    def get_output(self, worker_metadatas, buildroot_id):
         """
         Build the output entry of the metadata.
 
         :return: list, containing dicts of partial metadata
         """
         outputs = []
+        output_file = None
+
+        if self.source_build:
+            arch = os.uname()[4]
+            registry = self.workflow.push_conf.docker_registries[0]
+            build_name = get_unique_images(self.workflow)[0]
+            config = copy.deepcopy(registry.config)
+
+            # We don't need container_config section
+            if config and 'container_config' in config:
+                del config['container_config']
+
+            tags = set(image.tag for image in self.workflow.tag_conf.unique_images)
+
+            repositories = []
+            for reg in self.workflow.push_conf.all_registries:
+                new_repo = copy.deepcopy(build_name)
+                new_repo.registry = reg
+                repositories.append(new_repo)
+                pull_spec = copy.deepcopy(new_repo)
+
+                new_repo = copy.deepcopy(new_repo)
+                new_repo.tag = registry.digests[build_name.to_str()]
+                repositories.append(new_repo)
+
+            # need to pull image in order to get image_id
+            # or skopeo copy oci:/..... docker-daemon:someimage:tag
+            # will be available as docker.io/some:tag
+            self.tasker.pull_image(pull_spec)
+            image_inspect = self.tasker.inspect_image(pull_spec)
+            image_id = image_inspect['Id']
+
+            history = self.tasker.get_image_history(self.builder.image_id)
+            diff_ids = image_inspect[INSPECT_ROOTFS][INSPECT_ROOTFS_LAYERS]
+
+            # diff_ids is ordered oldest first
+            # history is ordered newest first
+            # We want layer_sizes to be ordered oldest first
+            layer_sizes = [{"diff_id": diff_id, "size": layer['Size']}
+                           for (diff_id, layer) in zip(diff_ids, reversed(history))]
+
+            self.tasker.remove_image(pull_spec, force=True)
+
+            metadata, output_file = self.get_image_output(arch, image_id, buildroot_id)
+            metadata.update({
+                'arch': arch,
+                'type': 'docker-image',
+                'components': [],
+                'extra': {
+                    'image': {
+                        'arch': arch,
+                    },
+                    'docker': {
+                        'id': image_id,
+                        'repositories': repositories,
+                        'layer_sizes': layer_sizes,
+                        'tags': list(tags),
+                        'config': config,
+                        'digests': registry.digests[build_name.to_str()],
+                    },
+                },
+            })
+            outputs.append(metadata)
```

Comment:
my reasoning was,
koji_upload serves primary for saving things for workers because there isn't any other way

since source container build doesn't have worker, koji_import can just get all it needs

but if you strongly feel that we should run koji_upload I'll move things there

Reply:
I'll make get_output general and use it in koji_upload and source part of koji_import